### ISSUE TITLE
Completion take long time to return

### ISSUE THREAD

**Issue by ChuckJonas** [neowit/tooling-force.com#18]
Hey,

I'm trying to leverage this jar to create a completion extension for vscode. 

I've succesfull been able to get the command to run on a project (mavensmate project):

```
java -jar tooling-force.jar --action=listCompletions --projectPath=/Users/johndoe/proj/ --currentFilePath=/Users/johndoe/proj/src/classes/FooBar.cls --currentFileContentPath=/Users/johndoe/proj/src/classes/FooBar.cls --line=15 --column=13 --responseFilePath=response.json --sf.username=johndoe@example.com --sf.password=pass+st --sf.serverurl=https://test.salesforce.com
```

However, it takes around 20 seconds to return a result.  Is there someway to cache the meta-data?  Mavensmate the symbol tables in it's config directory.

**Comment by ChuckJonas** [comment]
also, likely due to my configuration, this command only see to return anything if I run it on an SObject reference like so: 
``` java
Account acc = new Account();
acc.
```

For classes or even system methods it returns much faster but the result just contains `RESULT=SUCCESS`.  My guess would be it's failing to find the symbol files?

**Comment by ChuckJonas** [comment]
Sorry for spamming,

Completion on system namespaces (system,string,test) and custom classes does seem to return results. (within a second with is pretty quick).

Is there anything I can do to speed it up for SObject completion?  Maybe it's because I don't have the objects loaded in my workspace `src/object/`?

**Comment by neowit** [comment]
To take advantage of code completion it is best to run `tooling-force.jar` in "server" mode.

With your current setup (every time you call tooling-force.jar) you have to 
1. wait couple of seconds for JVM to start
2. tooling-force.jar to load in RAM. 
3. tooling-force.jar has to fetch specified SObject metadata 
or 
4. scan local classes to build AST.

When you run tooling-force.jar the way you do currently (non server mode) this process is repeated **every time**.  
In "server" mode however at least 3 (out of 4 steps above) do not have to happen on subsequent calls to `tooling-force.jar` because information is cached in RAM. It will still have to fetch new metadata if you try to call completion for a different SObject, but that result is also cached and ready for subsequent use.

Non server mode is more suitable for Continuous Integration scenarios and not for "IDE" like ones.

To start a server use `--action=serverStart` like so
```
java -Dfile.encoding=UTF-8  -jar /tmp/tooling-force.com.jar --action=serverStart --port=65000 --timeoutSec=1800

```
server start needs to be done only once (until server is unloaded based on `--timeoutSec` parameter).

To call the server use a TCP client (command line or from your favourite programming language).
Here is a command line example with [netcat](https://en.wikipedia.org/wiki/Netcat):
```
echo "--action=listCompletions --tempFolderPath='/tmp/apex/deployment' --config='/tmp/SFDC Credentials/MyProject.properties' --projectPath='/tmp/MyProject' --column=13 --line=11 --currentFilePath='/tmp/MyProject/src/classes/A_Fake_Class.cls' --currentFileContentPath='/tmp/MyProject/src/classes/A_Fake_Class.cls' --responseFilePath='/tmp/response_listCompletions' --pollWaitMillis=1000 --maxPollRequests=1000" | nc 127.0.0.1 65000
```

and here is an [example](https://github.com/neowit/vim-force.com/blob/legacy-vim-sync/apex-plugin/apexTooling.vim#L1328) of using python for the same purpose.

**Comment by ChuckJonas** [comment]
Awesome, 

Running in server mode makes sense (even though I was getting return times under a second for non-sobject auto-completes, it's probably because I'm on a new mbp).

Any reason you always retrieve the SObject info from SF, instead of first checking if the object exists in `src/objects`?

**Comment by ChuckJonas** [comment]
^ the reason I ask about the objects, is I would like to avoid using the username/password flow if possible.  (ideally i wouldn't need to auth the user for just code completion)

**Comment by neowit** [comment]
> Any reason you always retrieve the SObject info from SF, instead of first checking if the object exists in src/objects?

Because `src/objects`
A. may be outdated
B. not necessarily contain all fields of specific object (depends on package.xml )
C. does not provide good enough parent relationships and related child object details
D. contains a lot of information we do not need for completion but still would have to parse
E. SFDC keep adding new stuff into `.object` XML all the time, which would require updates of XML parser without any added value.

> ideally i wouldn't need to auth the user for just code completion

If I am not mistaken normal code completion (based only on local apex classes) does not require authentication. However as soon as you touch anything SObject or SOQL related there is no way to avoid calling SFDC because it is the only source of truth for this kind of data.

**Comment by ChuckJonas** [comment]
very good points...

Code completion for class does seem to work well without auth.

Is it possible to instead pass a access token (from oauth2 flow) instead of username and password?

**Comment by neowit** [comment]
> Is it possible to instead pass a access token (from oauth2 flow) instead of username and password?

Right now it is not supported, but I can look into this. Please open another ticket, as it is off topic for this one.

**Comment by ChuckJonas** [comment]
@neowit in service mode, can it handle multiple workspaces?  

EG: If i have multiple projects open will it properly be able to distinguish between the two?  Or do I need to spin up a service for each work space?

**Comment by neowit** [comment]
You should not need more than 1 server, regardless of the number of projects you are working with.

> If i have multiple projects open will it properly be able to distinguish between the two?

Yes, as long as you provide correct `--projectPath` value.
Each (project related) command you send to server must include project/workspace location (`--projectPath`), otherwise it will be refused.
